tests/unit/s2n_cipher_suite_match_test.c

```diff
@@ -781,6 +781,320 @@ int main(int argc, char **argv)
+                    &s2n_dhe_rsa_with_chacha20_poly1305_sha256,     /* 0xCC,0xAA */
+                    &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256,       /* 0xC0,0x23 */
+                    &s2n_ecdhe_ecdsa_with_chacha20_poly1305_sha256, /* 0xCC,0xA9 */
+                    &s2n_ecdhe_rsa_with_chacha20_poly1305_sha256,   /* 0xCC,0xA8 */
+                };
+
+                cipher_preferences = (struct s2n_cipher_preferences) {
+                    .count = s2n_array_len(cipher_suites_with_tls12_chacha20),
+                    .suites = cipher_suites_with_tls12_chacha20,
+                    .allow_chacha20_boosting = true
+                };
+
+                /* Server and client have chacha20 boosting enabled. Server is unable to negotiate any chacha20 cipher
+                 * suites, including its most preferred cipher suite.  
+                 */
+                {
+                    uint8_t dhe_chacha20_boosted_wire[] = {
+                        TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+                        TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+                        TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,
+                    };
+
+                    uint8_t count = sizeof(dhe_chacha20_boosted_wire) / S2N_TLS_CIPHER_SUITE_LEN;
+                    EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(connection, dhe_chacha20_boosted_wire, count));
+                    EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256);
+                }
+            }
+        }
+
+        /* Test ChaCha20 boosting behaviour for tls 1.3 connection */
+        {
+            EXPECT_SUCCESS(s2n_enable_tls13_in_test());
+
+            struct s2n_cipher_preferences cipher_preferences = { 0 };
+
+            struct s2n_security_policy security_policy = {
+                .minimum_protocol_version = S2N_SSLv3,
+                .cipher_preferences = &cipher_preferences,
+                .kem_preferences = &kem_preferences_null,
+                .signature_preferences = &s2n_signature_preferences_20201021,
+                .ecc_preferences = &s2n_ecc_preferences_test_all,
+            };
+    
+            DEFER_CLEANUP(struct s2n_connection *connection = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(connection);
+
+            connection->security_policy_override = &security_policy;
+            connection->kex_params.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+            connection->client_protocol_version = S2N_TLS13;
+            connection->actual_protocol_version = S2N_TLS13;
+            connection->server_protocol_version = S2N_TLS13;
+
+            /* The server chose a (malformed) security policy where ChaCha20 boosting is toggled on but 
+             * the cipher preferences does not contain ChaCha20. In this case, ChaCha20 boosting behaviour is off.
+             */
```

Comment:
We should probably test that the logic you added to s2n_security_policies prevents this case.